modules/core/05-port/types/module.go

```diff
@@ -82,4 +82,14 @@ type IBCModule interface {
 		packet channeltypes.Packet,
 		relayer sdk.AccAddress,
 	) error
+
+	// NegotiateAppVersion performs application version negotiation given the provided port ID, counterparty and proposed version.
+	// An error is returned if version negotiation cannot be performed. For example, an application module implementing this interface
+	// may decide to return an error in the event of the proposed version being incompatible with it's own
+	NegotiateAppVersion(
+		ctx sdk.Context,
+		portID string,
+		counterparty channeltypes.Counterparty,
+		proposedVersion string,
+	) (version string, err error)
```

Comment:
this still needs to take in the channel order and connection id

Reply:
Sure, I'll add these fields to `QueryAppVersionRequest` also 👍

Reply:
I'm in favor of `connectionID` with `string`. Connection hops can be confusing and we aren't even sure if multi hop will ever be baked into core IBC. If it is, it'll probably be API breaking anyways so better to force app developers to realize the change has been made

Reply:
Cool, thank you ❤️